servant/src/Servant/API/Verbs.hs

```diff
@@ -0,0 +1,166 @@
+-- <https://tools.ietf.org/html/rfc7231#section-6 RFC7231 Section 6>;
+-- the relevant information is summarily presented here.
+
+-- | 'GET' with 200 status code.
+type Get    contentTypes a = Verb 'GET    200 contentTypes a
+-- | 'POST' with 200 status code.
+type Post   contentTypes a = Verb 'POST   200 contentTypes a
+-- | 'PUT' with 200 status code.
+type Put    contentTypes a = Verb 'PUT    200 contentTypes a
+-- | 'DELETE' with 200 status code.
+type Delete contentTypes a = Verb 'DELETE 200 contentTypes a
+-- | 'PATCH' with 200 status code.
+type Patch  contentTypes a = Verb 'PATCH  200 contentTypes a
+
+-- * Other responses
+
+-- ** 201 Created
+--
+-- Indicates that a new resource has been created. The URI corresponding to the
+-- resource should be given in the @Location@ header field.
+--
+-- If the resource cannot be created immediately, use 'PostAccepted'.
+--
+-- Consider using 'Servant.Utils.Links.safeLink' for the @Location@ header
+-- field.
+
+-- | 'POST' with 201 status code.
+--
+type Created contentTypes a = Verb 'POST 201 contentTypes a
+
+
+-- ** 202 Accepted
+--
+-- Indicates that the request has been accepted for processing, but the
+-- processing has not yet completed. The status of the processing should be
+-- included, as well as either a link to a status monitoring endpoint or an
+-- estimate of when the processing will be finished.
+
+-- | 'GET' with 202 status code.
+type GetAccepted contentTypes a = Verb 'GET 202 contentTypes a
+-- | 'POST' with 202 status code.
+type PostAccepted contentTypes a = Verb 'POST 202 contentTypes a
+-- | 'DELETE' with 202 status code.
+type DeleteAccepted contentTypes a = Verb 'DELETE 202 contentTypes a
+-- | 'PATCH' with 202 status code.
+type PatchAccepted contentTypes a = Verb 'PATCH 202 contentTypes a
+-- | 'PUT' with 202 status code.
+type PutAccepted contentTypes a = Verb 'PUT 202 contentTypes a
+
+
+-- ** 203 Non-Authoritative Information
+--
+-- Indicates that the request has been successfully processed, but the
+-- information may come from a third-party.
+
+-- | 'GET' with 203 status code.
+type GetNonAuthoritative contentTypes a = Verb 'GET 203 contentTypes a
+-- | 'POST' with 203 status code.
+type PostNonAuthoritative contentTypes a = Verb 'POST 203 contentTypes a
+-- | 'DELETE' with 203 status code.
+type DeleteNonAuthoritative contentTypes a = Verb 'DELETE 203 contentTypes a
+-- | 'PATCH' with 203 status code.
+type PatchNonAuthoritative contentTypes a = Verb 'PATCH 203 contentTypes a
+-- | 'PUT' with 203 status code.
+type PutNonAuthoritative contentTypes a = Verb 'PUT 203 contentTypes a
+
+
+-- ** 204 No Content
+--
+-- Indicates that no response body is being returned. Handlers for these should
+-- return 'NoContent', possibly with headers.
+--
+-- If the document view should be reset, use @205 Reset Content@.
+
+-- | 'GET' with 204 status code.
+type GetNoContent contentTypes noContent = Verb 'GET 204 contentTypes noContent
```

Comment:
It was like, but @jkarni realized it disallowed to easily add headers to `NoContent` then.
With explicit `noContent` argument it is possible to add headers like this:

```haskell
GetNoContent '[JSON] (Headers '[Header "X-Custom-Header" CustomData] NoContent)
```